### DISCUSSION TITLE
Re-usable data

### DISCUSSION THREAD

**Discussion by desvu** [alpinejs/alpine#2555] [1. Help] [answered]
Hi! I am trying to create a component, but i am having some problems.

```html
<div x-data="test">
    <input type="text" x-bind="field">
</div>

<script>
    Alpine.data('test', () => ({
        str: '',
        // ...

        ['@resize.window']() { // this event is not binding with root element
            // ...
        },

        field: {
            ['x-model']() {
                return this.str // this code returns an empty string instead of a variable.
            },

            // ...
        },

        // ...
    }))
</script>
```

**Comment by SimoTod** [comment, marked as answer]
Events such as `@resize` need to be defined on the dom. I think you are mixing up Alpine data and x-bind.
Str is empty so if you never update it, it stays empty. Your x-bind is wrong x-model needs to return the name of the variable, not the value.

Your component should look like this one

```html
<div x-data="test">
  <input type="text" x-bind="field" @resize.window.debounce="foo()">
</div>

<script>
  Alpine.data('test', () => ({
    str: '',
    // ...
    foo() {
      // ...
    },
    field: {
      ['x-model']() {
        return 'str'
      },
      // ...
    },
    // ...
  }))
</script>
```

**Reply by desvu** [reply]
```html
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('test', () => ({
            str: 'Hello!',
          
            field: {
                ['x-model']() {
                  return 'str'
                }
            }
        }))
    })
</script>

<div x-data="test">
    <input type="text" x-bind="field">
    <!-- ... -->
</div>
```
![change](https://user-images.githubusercontent.com/16544732/147870366-de71e9d8-3ce1-4ac1-a60f-c22ac3e3f4c9.png)

**Reply by SimoTod** [reply]
> Not working :(
> 
> ```html
> <script>
>     document.addEventListener('alpine:init', () => {
>         Alpine.data('test', () => ({
>             str: 'Hello!',
> 
>             button: {
>                 ['@click']() {
>                     console.log('click!');
>                     this.str = 'World!';
>                 }
>             },
> 
>             field: {
>                 ['x-model']() {
>                     return 'str'
>                 }
>             }
>         }))
>     })
> </script>
> 
> <div x-data="test">
>     <input type="text" x-bind="field">
>   
>     <button type="button" x-bind="button">change</button>
> </div>
> ```
> ![change](https://user-images.githubusercontent.com/16544732/147869471-866f31a6-7c09-4339-b3e8-4a1ac1ca4416.png)
> 
>

It looks like x-model acts a bit weird with functions. Use `'x-model': 'str'` instead.

About the other question, you can put x-bind on the root element. You made the same mistake of the first example: defining a property starting with `@` in Alpine.data doesn't do anything by itself.

You are mixing up a few concepts, maybe start with some tutorial and simpler example to get familiar with all directives then move to more complex components.

**Reply by desvu** [reply]
Thanks!
The separate use of the x-model creates a big problem, in this case i need to remember the name of the component and the name of the variable name for the x-model.
